File: Advanced Programming Methods/example3/src/sample/model/ReadFile.java
package sample.model;

import sample.utils.FileData;
import sample.utils.MyIDictionary;
import sample.utils.exceptions.DivideByZeroException;
import sample.utils.exceptions.EmptyException;
import sample.utils.exceptions.NotExistingException;
import sample.utils.exceptions.OperatorException;

import java.io.BufferedReader;
import java.io.IOException;

public class ReadFile implements IStmt {
    private Exp exp_file_id;  //with this we get the id from symtable
    private String var_name;

    public ReadFile(Exp exp_file_id, String var_name){
        this.exp_file_id=exp_file_id;
        this.var_name=var_name;
    }

    @Override
    public PrgState execute(PrgState state) throws DivideByZeroException, NotExistingException, OperatorException, EmptyException, IOException {
        MyIDictionary<String, Integer> st=state.getSymTable();
        int id=exp_file_id.eval(state.getSymTable(), state.getHeap());
        FileData fd=state.getFileTable().get(id);
        BufferedReader bf=fd.getFileDescriptor();
        String s=bf.readLine();
        int number;
        if(s.equals(""))
            number=0;
        else
            number=Integer.parseInt(s);

        if(st.contains(var_name))
            st.update(var_name, number);
        else
            st.add(var_name, number);
        return state;
    }

    @Override
    public String toString(){
        return "ReadFile "+exp_file_id+" "+var_name;
    }
}

File: Advanced Programming Methods/toyLanguage/toyLanguage/model/ReadFile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;

namespace toyLanguage.model
{
    class ReadFile:IStmt
    {
        private Exp expFileId;
        private string varName;
        public ReadFile(Exp e,string vn)
        {
            varName = vn;
            expFileId = e;
        }
        public string VarName
        {
            get { return varName; }
            set { varName = value; }
        }

        public Exp ExpFileId
        {
            get { return expFileId; }
            set { expFileId = value; }
        }

        public PrgState Execute(PrgState p)
        {
            IMyDictionary<String, int> symbolT = p.SymTable;
            int id = expFileId.Eval(symbolT);
            FileData fd;
            p.FileTable.TryGetValue(id, out fd);
            StreamReader bf = fd.FileDescriptor;
            String s = bf.ReadLine();
            int number;
            if (s.Equals(""))
                number = 0;
            else
                Int32.TryParse(s, out number);

            if (symbolT.Exists(varName))
                symbolT.Update(varName, number);
            else
                symbolT.Add(varName, number);
            return p;

        }

        public override string ToString()
        {
            return "( ReadFile(" + this.expFileId + "," + this.varName + "))";
        }

    }
}
